storage/src/main/java/org/apache/kafka/storage/log/OffsetResultHolder.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.storage.log;
+
+import org.apache.kafka.common.errors.ApiException;
+import org.apache.kafka.common.record.FileRecords.TimestampAndOffset;
+
+import java.util.Objects;
+import java.util.Optional;
+
+public class OffsetResultHolder {
+
+    private Optional<TimestampAndOffset> timestampAndOffsetOpt;
+    private Optional<AsyncOffsetReadFutureHolder<FileRecordsOrError>> futureHolderOpt;
+    private Optional<ApiException> maybeOffsetsError = Optional.empty();
+    private Optional<Long> lastFetchableOffset = Optional.empty();
+
+    public OffsetResultHolder() {
+        this(Optional.empty(), Optional.empty());
+    }
+
+    public OffsetResultHolder(
+            Optional<TimestampAndOffset> timestampAndOffsetOpt,
+            Optional<AsyncOffsetReadFutureHolder<FileRecordsOrError>> futureHolderOpt
+    ) {
+        this.timestampAndOffsetOpt = timestampAndOffsetOpt;
+        this.futureHolderOpt = futureHolderOpt;
+    }
+
+    public OffsetResultHolder(Optional<TimestampAndOffset> timestampAndOffsetOpt) {
```

Comment:
Is there a downside to defining a few more constructors which take the concrete type and create an Optional internally? I believe this will save you a lot of `Optional.of(new Something()` elsewhere, no?